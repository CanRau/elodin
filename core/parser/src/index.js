import Parser from './Parser'
import parse from './parse'
import errorTypes from './errorTypes'

export { Parser, parse, errorTypes }
