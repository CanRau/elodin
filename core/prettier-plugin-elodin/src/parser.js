import { format } from '@elodin/format'

export default function parse(text, _, options) {
  const formatted = format(text)

  const errors = formatted.errors
  let formattedText = formatted.code

  if (errors.length > 0) {
    let formattedText = text
  }

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    error: errors[0],
    end: text.length,
    source: text,
    start: 0,
  }
}
