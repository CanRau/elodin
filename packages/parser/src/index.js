import Parser from './Parser'
import parse from './parse'

export { Parser, parse }
