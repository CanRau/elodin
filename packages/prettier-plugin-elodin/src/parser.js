import { format } from '@elodin/format'

export default function parse(text) {
  const formattedText = format(text)

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    end: text.length,
    source: text,
    start: 0,
  }
}
