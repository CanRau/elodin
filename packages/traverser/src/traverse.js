import traverseNode from './traverseNode'

export default function traverse(ast, visitors = []) {
  visitors.forEach(visitor => traverseNode(ast, visitor))

  return ast
}
