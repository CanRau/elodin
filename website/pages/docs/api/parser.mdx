import { make as DocLayout } from '../../../components/DocLayout.bs.js'

export default DocLayout

# Parser

`@elodin/parser`

The parser turns a string of Elodin code into an AST (Abstract Syntax Tree).

## parse

It exposes a `parse` method that just takes an single string of code.

### Example

```javascript
import { parse } from '@elodin/parser'

const input = `
view Button {
  backgroundColor: red
  paddingLeft: 10
  paddingRight: 10
}
`

const ast = parse(input)
```

## AST Nodes

The AST consists of the following AST nodes:

### File

| Property | Type | Description |
| -------- | ---- | ----------- |
| body     |      |             |

### Style

| Property | Type | Description |
| -------- | ---- | ----------- |
| name     |      |             |
| format   |      |             |
| body     |      |             |
| comments |      |             |

### Variant

| Property | Type | Description |
| -------- | ---- | ----------- |
| name     |      |             |
| body     |      |             |
| comments |      |             |

### Conditional

| Property | Type | Description |
| -------- | ---- | ----------- |
| property |      |             |
| value    |      |             |
| operator |      |             |
| boolean  |      |             |
| body     |      |             |

### Declaration

| Property | Type | Description |
| -------- | ---- | ----------- |
| property |      |             |
| value    |      |             |
| dynamic  |      |             |
| raw      |      |             |

### FunctionExpression

| Property | Type | Description |
| -------- | ---- | ----------- |
| callee   |      |             |
| params   |      |             |

### Float

| Property   | Type | Description |
| ---------- | ---- | ----------- |
| integer    |      |             |
| fractional |      |             |
| negative   |      |             |

### Integer

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
| negative |      |             |

### Identifier

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

### String

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |

### Variable

| Property | Type | Description |
| -------- | ---- | ----------- |
| value    |      |             |
