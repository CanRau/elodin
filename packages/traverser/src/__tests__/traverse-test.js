import { parse } from '@elodin/parser'

import traverse from '../traverse'

describe('Traversing ast nodes', () => {
  it('should correctly traverse nodes', () => {
    const file = `
style Button {
  color: red
  [Type=Primary] {
    color: blue
  }
}`

    const log = []

    const visitor = {
      Identifier: {
        enter(node) {
          log.push(node.value)
          node.value = node.value.toUpperCase()
          log.push(node.value)
        },

        exit(node) {
          log.push(node.value)
        },
      },
    }

    const { ast } = parse(file)

    expect(traverse(ast, [visitor])).toMatchSnapshot()
    expect(log).toMatchSnapshot()
  })
})
