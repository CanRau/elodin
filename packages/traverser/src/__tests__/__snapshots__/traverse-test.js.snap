// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traversing ast nodes should correctly traverse nodes 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "property": "color",
          "type": "Declaration",
          "value": Object {
            "type": "Identifier",
            "value": "RED",
          },
        },
        Object {
          "body": Array [
            Object {
              "property": "color",
              "type": "Declaration",
              "value": Object {
                "type": "Identifier",
                "value": "blue",
              },
            },
          ],
          "operator": "=",
          "property": Object {
            "type": "Identifier",
            "value": "TYPE",
          },
          "type": "Conditional",
          "value": Object {
            "type": "Identifier",
            "value": "PRIMARY",
          },
        },
      ],
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;

exports[`Traversing ast nodes should correctly traverse nodes 2`] = `
Array [
  "red",
  "RED",
  "RED",
  "Type",
  "TYPE",
  "TYPE",
  "Primary",
  "PRIMARY",
  "PRIMARY",
  "blue",
  "blue",
  "blue",
]
`;
